ecl/hql/hqlfold.cpp

```diff
@@ -1335,8 +1346,486 @@ IValue * foldExternalCall(IHqlExpression* expr, unsigned foldOptions, ITemplateC
+
+    virtual unsigned logString(const char *text) const { throwUnexpected(); }
+    virtual IDebuggableContext *queryDebugContext() const { return NULL; }
+
+    // Memory management
+
+    virtual IEngineRowAllocator * getRowAllocator(IOutputMetaData * meta, unsigned activityId) const { throwUnexpected(); }
+    virtual const char * cloneVString(const char *str) const { throwUnexpected(); }
+    virtual const char * cloneVString(size32_t len, const char *str) const { throwUnexpected(); }
+
+    // Called from generated code for FROMXML/TOXML
+
+    virtual const void * fromXml(IEngineRowAllocator * _rowAllocator, size32_t len, const char * utf8, IXmlToRowTransformer * xmlTransformer, bool stripWhitespace) { throwUnexpected(); }
+    virtual void getRowXML(size32_t & lenResult, char * & result, IOutputMetaData & info, const void * row, unsigned flags) { throwUnexpected(); }
+
+    // Miscellaneous
+
+    virtual void getExternalResultRaw(unsigned & tlen, void * & tgt, const char * wuid, const char * stepname, unsigned sequence, IXmlToRowTransformer * xmlTransformer, ICsvToRowTransformer * csvTransformer) { throwUnexpected(); }    // shouldn't really be here, but it broke thor.
+    virtual char * queryIndexMetaData(char const * lfn, char const * xpath) { throwUnexpected(); }
+
+    // Called from generated code for FROMJSON
+
+    virtual const void * fromJson(IEngineRowAllocator * _rowAllocator, size32_t len, const char * utf8, IXmlToRowTransformer * xmlTransformer, bool stripWhitespace) { throwUnexpected(); }
+    virtual void getRowJSON(size32_t & lenResult, char * & result, IOutputMetaData & info, const void * row, unsigned flags) { throwUnexpected(); }
+
+    virtual const IContextLogger &queryContextLogger() const
+    {
+        return queryDummyContextLogger();
+    }
+    virtual IEngineContext *queryEngineContext() { return NULL; }
+    virtual char *getDaliServers() { throwUnexpected(); }
+    virtual IWorkUnit* updateWorkUnit() const { throwUnexpected(); }
+    virtual ISectionTimer * registerTimer(unsigned activityId, const char * name) { throwUnexpected(); }
+    virtual IEngineRowAllocator * getRowAllocatorEx(IOutputMetaData * meta, unsigned activityId, unsigned flags) const { throwUnexpected(); }
+
+};
+
+IHqlExpression *deserializeConstantSet(ITypeInfo *type, bool isAll, size32_t len, const void *vdata)
+{
+    ITypeInfo *childType = type->queryChildType();
+    if (isAll)
+        return createValue(no_all, LINK(childType));
+    else if (!len)
+        return createValue(no_null, LINK(childType));
+    else
+    {
+        const char *data = (const char *) vdata;
+        const char *end = data + len;
+        HqlExprArray values;
+        while (data < end)
+        {
+            size32_t size = childType->getSize();
+            size32_t numChars = childType->getStringLen();
+            switch (childType->getTypeCode())
+            {
+            case type_int:
+                if (childType->isSigned())
```

Comment:
They could but createValueFromMem would also need to update the pointer according to how much data was being read... I don't think it would be a net gain in clarity/maintainability. I can call it for int/real/bool though...